// blockchain-ui-privatization/PC/common-mixin/contract/index/index.js
import { mapState } from 'vuex';
import { myStorage } from '@/utils';
import worker from '@/utils/webWorker';

export default {
  name: 'trade',
  data() {
    return {
      // 是否是收起状态
      shrink: false,
      // 24小时行情 是否是收起状态
      marketShrink: false,
      // 窗口宽度
      screenWidth: null,
      // 延迟执行
      windowOnResizeFire: true,
      // 宽度分割
      mediaWidth: 1820,
      //
      coPublicInfo: null,
      // shuju
      marketData: null,
      // MywebSocket
      MywebSocket: null,
      // 当前市场
      marketCurrent: null,
      // 当前币对
      symbolCurrent: null,
      // 深度级别
      depthClasses: '0',
      // 深度值
      depthValue: 0,
      // 图表遮罩
      coverKlineBox: null,
      isFull: null,
      routeSymbol: this.$route.params.symbol,
    };
  },
  computed: {
    ...mapState({
      baseInfo({ baseData }) {
        if (baseData.coPublicInfo) {
          this.coPublicInfo = baseData.coPublicInfo;
          this.marketData = this.coPublicInfo.market;
          // 获取当前市场
          this.marketCurrent = myStorage.get('coMarkTitle');
          // 获取当前币对
          this.symbolCurrent = myStorage.get('coNowSymbol');
          return baseData.coPublicInfo;
        }
        return null;
      },
    }),
    templateLayoutType() {
      return this.$store.state.baseData.templateLayoutType;
    },
    worker() {
      return worker();
    },
    // 是否Login
    isLogin() {
      return this.$store.state.baseData.isLogin;
    },
    // 汇率单位
    rateData() {
      return this.$store.state.baseData.rate;
    },
    // 默认 获取 所有币对的
    symbolAll() {
      if (this.marketData) {
        let arr = [];
        const obj = {};
        const marketKeys = Object.keys(this.marketData);
        marketKeys.forEach((item) => {
          arr = arr.concat(this.marketData[item]);
        });
        arr.forEach((k) => {
          obj[k.symbol] = k;
          obj[k.symbol].name = k.symbol;
          obj.tradeType = 'contract';
        });
        this.$bus.$emit('SYMBOL_LIST_ALL', obj);
        return obj;
      }
      return null;
    },
    symbolCurrentInfo() {
      if (this.symbolAll && this.symbolCurrent) {
        return this.symbolAll[this.symbolCurrent];
      }
      return null;
    },
  },
  watch: {
    // 监听 获取到数据
    coPublicInfo(val) {
      if (val.wsUrl) {
        // 创建WS
        this.worker.postMessage({
          type: 'CREAT_WEBSOCKET',
          data: {
            wsUrl: val.wsUrl,
            lan: this.$store.state.baseData.lan,
            rate: this.rateData,
            symbolAll: this.symbolAll,
          },
        });
      }
    },
    // 监听 当前选中货币对的改变
    symbolCurrent(val, oldVla) {
      if (val) {
        const sSymbol = val.replace('/', '_');
        this.$router.push(`/trade/${sSymbol}`);
        this.$bus.$emit('SYMBOL_CURRENT', val);
      }
      if (oldVla && this.MywebSocket) {
        // 停止 上一个币对的 实时成交 send
        this.webSocketSend('Trade', 'unsub', oldVla);
        // 发送 盘口深度数据 Send
        this.webSocketSend('Depth', 'unsub', oldVla, '0');
        // 发送 实时成交 数据 Send
        this.webSocketSend('Trade', 'req', val);
        this.webSocketSend('Trade', 'sub', val);
        // 发送 盘口深度 数据 Send
        this.webSocketSend('Depth', 'sub', val, '0');
      }
    },
    // 监听 webSocket 创建成功
    MywebSocket(val) {
      if (val) {
        // 发送 24小时行情历史数据 Send
        this.webSocketSend('Market', 'sub', this.symbolCurrent, this.symbolAll);
        // // 发送 实时成交 数据 Send
        this.webSocketSend('Trade', 'req', this.symbolCurrent);
        this.webSocketSend('Trade', 'sub', this.symbolCurrent);
        // // 发送 盘口深度数据 Send
        this.webSocketSend('Depth', 'sub', this.symbolCurrent, 0);
      }
    },
    // 当前币对的信息数据
    symbolCurrentInfo(val) {
      this.$bus.$emit('SYMBOL_CURRENT_INFO', val);
    },
  },
  methods: {
    init() {
      if (this.routeSymbol) {
        myStorage.set('coMarkTitle', this.routeSymbol.split('_')[1]);
        myStorage.set('coNowSymbol', this.routeSymbol);
      } else {
        const sSymbolName = myStorage.get('sSymbolName');
        const sSymbol = sSymbolName;
        this.$router.push(`/trade/${sSymbol}`);
      }
      this.listenBusEvent();
      this.worker.onmessage = (event) => {
        const { data } = event;
        // 监听 WebSocket 链接成功
        if (data.type === 'WEBSOCKET_ON_OPEN') {
          this.MywebSocket = data.data.type;
          this.$bus.$emit('WEBSOCKET_ON_OPEN', this.MywebSocket);
        }
        // 监听 WS 数据
        if (data.type === 'WEBSOCKET_DATA') {
          this.listenWSData(data.data);
        }
      };

      this.screenWidth = document.body.clientWidth;
      if (this.screenWidth <= this.mediaWidth) {
        this.shrink = true;
        this.marketShrink = true;
      }
      // 监听 窗口的大小改变
      this.$bus.$on('WINFOW_ON_RESIIZE', () => {
        if (this.windowOnResizeFire) {
          this.screenWidth = document.body.clientWidth;
          if (this.screenWidth <= this.mediaWidth) {
            // 如果 屏幕窗口 小于阈值 设置成收起模式
            this.shrink = true;
            this.marketShrink = true;
          } else {
            // 如果 屏幕窗口 大于阈值 设置成展开模式
            this.shrink = false;
            this.marketShrink = false;
          }
          this.windowOnResizeFire = false;
          // 0.3秒之后将标志位重置
          setTimeout(() => {
            this.windowOnResizeFire = true;
          });
        }
      });
    },
    listenBusEvent() {
      // 监听 市场切换
      this.$bus.$on('ON_MARKET_SWITCH', (data) => {
        this.marketCurrent = data;
        myStorage.set('coMarkTitle', data);
      });
      // 监听 市场切换
      this.$bus.$on('ON_MARKET_SWITCH_ORDER', (data) => {
        this.marketCurrent = data;
        myStorage.set('coMarkTitle', data);
      });
      // 监听 币对切换
      this.$bus.$on('ON_SYMBOL_SWITCH', (data) => {
        myStorage.set('coNowSymbol', data);
        this.symbolCurrent = data;
      });
      // 监听 深度级别的切换
      this.$bus.$on('DEPTH_CLASSES', (data) => {
        this.depthClasses = data;
      });
      // 监听 深度级别的值
      this.$bus.$on('DEPTH_VALUE', (data) => {
        this.depthValue = data;
      });
      // 监听 kline 发送Send
      this.$bus.$on('WEBSOCKET_KLINE_SEND', (data) => {
        this.worker.postMessage({
          type: 'WEBSOCKET_KLINE_SEND',
          data,
        });
      });
    },
    listenWSData(data) {
      const { type, WsData } = data;
      this.$bus.$emit(type, WsData);
      if (type === 'TRADE_DATA') {
        this.worker.postMessage({
          type: 'TRADE_QUEUE_DATA',
          data: WsData.queueDataList,
        });
      }
    },
    webSocketSend(type, sendType, symbolData, symbolList) {
      this.worker.postMessage({
        type: 'WEBSOCKET_SEND',
        data: {
          type,
          sendType,
          symbolData,
          symbolList,
          depthValue: this.symbolCurrentInfo.pricePrecision,
        },
      });
    },
    ondblclick() {
      return false;
    },
    onclickfun(e) {
      // 显示和隐藏币币交易页面 TradingView 的遮罩
      if (e.target.className === 'coverKlineBox') {
        this.coverKlineBox = e.target;
      }
      if (e.target.className === 'coverKlineBox') {
        this.coverKlineBox.style.display = 'none';
      } else if (this.coverKlineBox) {
        this.coverKlineBox.style.display = 'block';
        this.coverKlineBox = null;
      }
    },
  },

};
